hbase-server/src/test/java/org/apache/hadoop/hbase/TestZooKeeper.java

```diff
@@ -176,7 +176,7 @@ private void testSanity(final String testName) throws Exception {
   @Test
   public void testRegionAssignmentAfterMasterRecoveryDueToZKExpiry() throws Exception {
     MiniHBaseCluster cluster = TEST_UTIL.getHBaseCluster();
-    cluster.startRegionServer();
+    cluster.startRegionServerAndWait(2000);
```

Comment:
I'm curious why these are needed in multiple places, hardcoding makes these tests flaky?

Reply:
`startRegionServer` and `startRegionServerAndWait` do the same thing except the latter waits up to the passed in timeout to poll for the rs's online status in master (it will return early if it finds the rs is online before the timeout). the `startRegionServerAndWait` method gives the master a bit more time to get the updated `online` status from regionserver (which is updated after `HRegionServer.handleReportForDutyResponse` but only reported to master a while later, in `HRegionServer.tryRegionServerReport`). that's why master often needs an extra second to know the rs is alive, and if we try to proceed to do other tasks without this wait (such as assign regions), master will run into errors because it doesn't know the rs is alive yet.

Reply:
There seems to be a race condition. Multiple tests rely on this guarantee that the regionserver is up by the time this call is returned. A fixed sleep doesn't seem like a right solution.  Instead we should fix the following method

``` 
  /**
     * Block until the region server has come online, indicating it is ready
     * to be used.
     */
    public void waitForServerOnline() {
      // The server is marked online after the init method completes inside of
      // the HRS#run method.  HRS#init can fail for whatever region.  In those
      // cases, we'll jump out of the run without setting online flag.  Check
      // stopRequested so we don't wait here a flag that will never be flipped.
      regionServer.waitForServerOnline();
    }
  }
```

That translates to when `online` is marked in HRegionServer. Currently it is marked 


```
 public void waitForServerOnline(){
    while (!isStopped() && !isOnline()) {
      synchronized (online) {
        try {
          online.wait(msgInterval);
        } catch (InterruptedException ie) {
          Thread.currentThread().interrupt();
          break;
        }
      }
    }
  }
```

Instead shouldn't it check when it is marked online by master (which happens after the first report is sent)? (block until the regionserver is marked online in the active master's server manager)

@saintstack WDYT? This doesn't cause test flakes?

Reply:
I think your advice solid.  I did not realize it a wait; it will be flakey on different platforms. I like the idea of checking online atomic

Reply:
@bharathv the `waitForServerOnline()` method is specific to `JVMClusterUtil.RegionServerThread`. I guess to your point, we could replace `t.waitForServerOnline()` with `getMaster().getServerManager().getOnlineServersList().contains(t);` inside `MiniHBaseCluster.startRegionServer`:

```
private JVMClusterUtil.RegionServerThread startRegionServer(Configuration configuration)
      throws IOException {
    User rsUser =
        HBaseTestingUtility.getDifferentUser(configuration, ".hfs."+index++);
    JVMClusterUtil.RegionServerThread t =  null;
    try {
      t = hbaseCluster.addRegionServer(
          configuration, hbaseCluster.getRegionServers().size(), rsUser);
      t.start();
      getMaster().getServerManager().getOnlineServersList().contains(t);
    } catch (InterruptedException ie) {
      throw new IOException("Interrupted adding regionserver to cluster", ie);
    }
    return t;
  }
```

would that resolve the issue?

Reply:
@bharathv if we want to get the online servers list from the active master in the `JVMClusterUtil.waitForServerOnline` method, wouldn't we have to change the signature of the method to pass in a list of masters?

Reply:
@caroliney14 please say more. I'm not sure I follow sir.

Reply:
@saintstack Its the other way around. After this patch, "online" atomic is not set to true at the right place. It is being set in handleReportForDutyResponse() which doesn't mean the RS is online. It should either be set after the first RS report or _ waitForServerOnline_ should check the active master's online list of masters.

I was suggesting the latter to avoid extra code refactoring. So yes, we should pass the active master object to it _or_ we could just switch to using HBaseCluster#waitForRegionServerToStart() that relies on the onlineServers from active master. Latter approach might be easier to refactor IIUC.

Reply:
@bharathv I'm not sure I follow -- `HBaseCluster.waitForRegionServerToStart` uses the same rs online checking logic as the `MiniHBaseCluster.startRegionServerAndWait` that is currently being used. Both query master's onlineServers list for the rs's online status (using `getClusterMetrics().getLiveServerMetrics().keySet()`) and both time out if the rs cannot be found within that list within the given timeout. Is there any difference to using `HBaseCluster.waitForRegionServerToStart`/`MiniHBaseCluster.startRegionServerAndWait` versus refactoring `waitForServerOnline` to query the master? If we refactored `waitForServerOnline` it would still have to use a polling mechanism with timeout, right?

Reply:
Thanks @bharathv  ... I'd think RS#online should change as part of this patch else we'll all be confused going forward around what 'online' means.

`and both time out if the rs cannot be found within that list within the given timeout. `

Why does it timeout do you think @caroliney14 ? Its not something like foreground thread holding up the RS startup?

(This is a nice PR)

Reply:
> and both time out if the rs cannot be found within that list within the given timeout.

@saintstack What I meant by this is, the behavior of the `HBaseCluster.waitForRegionServerToStart` and `MiniHBaseCluster.startRegionServerAndWait` methods are such that the methods themselves will time out and throw an exception if the rs is not found in master's online servers list within a specified time period. The method shouldn't time out if the master/rs communication logic is correct (when I tested locally, it doesn't). See method pasted below:

```
/**
   * Wait for the specified region server to join the cluster
   * @throws IOException if something goes wrong or timeout occurs
   */
  public void waitForRegionServerToStart(String hostname, int port, long timeout)
      throws IOException {
    long start = System.currentTimeMillis();
    while ((System.currentTimeMillis() - start) < timeout) {
      for (ServerName server : getClusterMetrics().getLiveServerMetrics().keySet()) {
        if (server.getHostname().equals(hostname) && server.getPort() == port) {
          return;
        }
      }
      Threads.sleep(100);
    }
    throw new IOException("did timeout " + timeout + "ms waiting for region server to start: "
        + hostname);
  }
```

As for changing where rs's `online` flag is set -- I guess it depends on what we wish to define as "online" for a regionserver. If "online" for a regionserver means that the regionserver has finished setup, then it's fine where it is. If "online" for a regionserver should mean that the regionserver has communicated to master it's ready to be assigned regions, then `online` should be set after the first `reportForDuty`.

Reply:
Sorry I was busy, just reading thru the comments.

> If "online" for a regionserver should mean that the regionserver has communicated to master it's ready to be assigned regions, then online should be set after the first reportForDuty.

This... Earlier (without your patch), "online" meant this ^^. Now with your patch, the semantics have changed, so what @saintstack is saying (IIUC) is to preserve those semantics by setting it after the first `reportForDuty` and then you don't have to make any changes to the tests (if this theory is correct).

Reply:
sorry, I meant to say "If online for a regionserver should mean that the regionserver has communicated to master it's ready to be assigned regions, then online should be set after the first `tryRegionServerReport`," since `online` is currently being set in `handleReportForDuty`, after `reportForDuty` finishes (see [this line](https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java#L1657-L1660)). if we agree on this semantic, then I'll make the change, within [this block](https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java#L1053-L1087). it would look something like this:

```
      // Run mode.
      long lastMsg = System.currentTimeMillis();
      long oldRequestCount = -1;
      boolean firstReport = true;
      // The main run loop.
      while (!isStopped() && isHealthy()) {
      ...
      long now = System.currentTimeMillis();
        if ((now - lastMsg) >= msgInterval) {
          // Register with the Master now that our setup is complete.
          tryRegionServerReport(lastMsg, now);
          if (firstReport) {
            // Wake up anyone waiting for this server to online
            synchronized (online) {
              online.set(true);
              online.notifyAll();
            }
            firstReport = false;
          }
          lastMsg = System.currentTimeMillis();
        }
        if (!isStopped() && !isAborted()) {
          this.sleeper.sleep();
        }
      ...
```

Reply:
You need do this only if tryRegionServerReport() is successful. It can also return without a successful report if the exception thrown is not `YouAreDeadException`. You may need to make that method return a boolean that tells the caller if the report was successfully sent.

Another nit: you don't need a separate boolean firstReport

```
if (!online.get()) {
  synchronized(online) {
    ....
  }
}
```

Also please undo the test changes along with this change to validate if they are passing now.

Reply:
Testing locally, this way seems to work. I've undone all changes to the tests except for two test files which actually have bugs.